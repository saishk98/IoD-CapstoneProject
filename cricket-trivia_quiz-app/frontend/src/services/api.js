import axios from "axios";

const BASE_URL = "http://localhost:5000/api/quiz";

export async function getCategories() {
  const response = await axios.get(`${BASE_URL}/categories`);
  return response.data;
}

export async function getQuestions(category) {
  console.log("getQuestions", category);
  const response = await axios.get(
    `${BASE_URL}/questions?category=${category}`
  );
  return response.data;
}

export async function submitAnswer(
  userId,
  category,
  selectedAnswer,
  correctAnswer,
  startTime
) {
  const response = await axios.post(`${BASE_URL}/submit`, {
    userId,
    category,
    selectedAnswer,
    correctAnswer,
    startTime,
  });
  return response.data;
}

export async function getLeaderboard() {
  const response = await axios.get(`${BASE_URL}/leaderboard`);
  return response.data;
}

export const saveScore = async (playerData) => {
    const response = await fetch("http://localhost:5000/api/leaderboard", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(playerData),
    });

    return response.json();
};
