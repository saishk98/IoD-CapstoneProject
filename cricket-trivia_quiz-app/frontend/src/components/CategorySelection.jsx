import React, { useEffect, useState } from "react";
import { getCategories } from "../services/api";
import { Button, Grid, CircularProgress } from "@mui/material";
import "../styles/category.css";

const CategorySelection = ({ setCategory }) => {
    const [categories, setCategories] = useState([]);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        getCategories().then(data => {
            setCategories(data);
            setLoading(false);
        });
    }, []);

    return (
        <div className="category-container">
            <h2 className="slide-in">Select a Quiz Category</h2>

            {loading ? (
                <CircularProgress color="primary" />
            ) : (
                <Grid container spacing={2} justifyContent="center">
                    {categories.map((category, index) => (
                        <Grid item key={index}>
                            <Button 
                                variant="contained" 
                                color="primary" 
                                onClick={() => setCategory(category)} 
                                className="bounce"
                            >
                                {category}
                            </Button>
                        </Grid>
                    ))}
                </Grid>
            )}
        </div>
    );
};

export default CategorySelection;