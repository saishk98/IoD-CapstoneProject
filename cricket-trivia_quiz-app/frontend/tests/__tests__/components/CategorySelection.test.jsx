import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import CategorySelection from "../../../src/components/CategorySelection";
import { getCategories } from "../../../src/services/api";
import { vi, describe, it, beforeEach, expect } from "vitest";

// Mock the getCategories API
vi.mock("../../../src/services/api", () => ({
    getCategories: vi.fn(),
}));

describe("CategorySelection", () => {
    const mockSetCategory = vi.fn();

    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders loading spinner initially", () => {
        getCategories.mockReturnValue(new Promise(() => {})); // never resolves
        render(<CategorySelection setCategory={mockSetCategory} />);
        expect(screen.getByRole("progressbar")).toBeInTheDocument();
        expect(screen.getByText(/Select a Quiz Category/i)).toBeInTheDocument();
    });

    it("renders categories after loading", async () => {
        getCategories.mockResolvedValue(["Cricket", "History", "Science"]);
        render(<CategorySelection setCategory={mockSetCategory} />);
        await waitFor(() => {
            expect(screen.getByText("Cricket")).toBeInTheDocument();
            expect(screen.getByText("History")).toBeInTheDocument();
            expect(screen.getByText("Science")).toBeInTheDocument();
        });
        expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
    });

    it("calls setCategory when a category button is clicked", async () => {
        getCategories.mockResolvedValue(["Cricket"]);
        render(<CategorySelection setCategory={mockSetCategory} />);
        const button = await screen.findByText("Cricket");
        fireEvent.click(button);
        expect(mockSetCategory).toHaveBeenCalledWith("Cricket");
    });

    it("renders no buttons if categories array is empty", async () => {
        getCategories.mockResolvedValue([]);
        render(<CategorySelection setCategory={mockSetCategory} />);
        await waitFor(() => {
            expect(screen.queryAllByRole("button")).toHaveLength(0);
        });
    });

    it("matches snapshot after loading", async () => {
        getCategories.mockResolvedValue(["Cricket", "History"]);
        const { asFragment } = render(<CategorySelection setCategory={mockSetCategory} />);
        await screen.findByText("Cricket");
        expect(asFragment()).toMatchSnapshot();
    });
});