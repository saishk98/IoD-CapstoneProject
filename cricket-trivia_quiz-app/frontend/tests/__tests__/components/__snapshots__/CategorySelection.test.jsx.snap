// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategorySelection > matches snapshot after loading 1`] = `
<DocumentFragment>
  <div
    class="category-container"
  >
    <h2
      class="slide-in"
    >
      Select a Quiz Category
    </h2>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-1ts58i8-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary bounce css-74d805-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cricket
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary bounce css-74d805-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          History
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
